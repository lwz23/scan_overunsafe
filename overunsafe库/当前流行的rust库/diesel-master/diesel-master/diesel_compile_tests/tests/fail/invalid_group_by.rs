extern crate diesel;

use diesel::alias;
use diesel::prelude::*;

table! {
    users {
        id -> Integer,
        name -> VarChar,
    }
}

table! {
    posts {
        id -> Integer,
    }
}

allow_tables_to_appear_in_same_query!(users, posts);
allow_columns_to_appear_in_same_group_by_clause!(users::id, posts::id);

fn main() {
    let conn = &mut PgConnection::establish("..").unwrap();

    // this fails because `posts` is not part of the from clause
    users::table
        .group_by(posts::id)
        .select(users::id)
        .execute(conn)
        .unwrap();

    // order of select and group by does not matter for the error
    users::table
        .select(users::id)
        .group_by(posts::id)
        .execute(conn)
        .unwrap();

    let (user_alias, post_alias) = alias!(users as user1, posts as post1,);

    // this also fails if we use aliases
    user_alias
        .group_by(posts::id)
        .select(user_alias.field(users::id))
        .execute(conn)
        .unwrap();

    users::table
        .group_by(post_alias.field(posts::id))
        .select(users::id)
        .execute(conn)
        .unwrap();

    user_alias
        .group_by(post_alias.field(posts::id))
        .select(user_alias.field(users::id))
        .execute(conn)
        .unwrap();

    user_alias
        .select(user_alias.field(users::id))
        .group_by(posts::id)
        .execute(conn)
        .unwrap();

    users::table
        .select(users::id)
        .group_by(post_alias.field(posts::id))
        .execute(conn)
        .unwrap();

    user_alias
        .select(user_alias.field(users::id))
        .group_by(post_alias.field(posts::id))
        .execute(conn)
        .unwrap();
}
