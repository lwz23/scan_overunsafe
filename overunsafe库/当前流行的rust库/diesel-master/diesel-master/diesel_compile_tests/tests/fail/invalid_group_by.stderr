error[E0271]: type mismatch resolving `<FromClause<table> as AppearsInFromClause<table>>::Count == Once`
  --> tests/fail/invalid_group_by.rs:27:10
   |
27 |         .group_by(posts::id)
   |          ^^^^^^^^ expected `Never`, found `Once`
   |
note: required for `posts::columns::id` to implement `AppearsOnTable<FromClause<users::table>>`
  --> tests/fail/invalid_group_by.rs:15:9
   |
15 |         id -> Integer,
   |         ^^
   = note: associated types for the current `impl` cannot be restricted in `where` clauses
   = note: required for `SelectStatement<FromClause<users::table>>` to implement `GroupByDsl<posts::columns::id>`

error[E0271]: type mismatch resolving `<id as IsContainedInGroupBy<id>>::Output == Yes`
  --> tests/fail/invalid_group_by.rs:28:10
   |
28 |         .select(users::id)
   |          ^^^^^^ type mismatch resolving `<id as IsContainedInGroupBy<id>>::Output == Yes`
   |
note: expected this to be `diesel::expression::is_contained_in_group_by::No`
  --> tests/fail/invalid_group_by.rs:15:9
   |
15 |         id -> Integer,
   |         ^^
note: required for `users::columns::id` to implement `ValidGrouping<posts::columns::id>`
  --> tests/fail/invalid_group_by.rs:8:9
   |
8  |         id -> Integer,
   |         ^^
   = note: associated types for the current `impl` cannot be restricted in `where` clauses
   = note: required for `SelectStatement<FromClause<table>, DefaultSelectClause<FromClause<table>>, ..., ..., ..., ..., ...>` to implement `SelectDsl<users::columns::id>`
   = note: consider using `--verbose` to print the full type name to the console

error[E0271]: type mismatch resolving `<FromClause<table> as AppearsInFromClause<table>>::Count == Once`
  --> tests/fail/invalid_group_by.rs:35:10
   |
35 |         .group_by(posts::id)
   |          ^^^^^^^^ expected `Never`, found `Once`
   |
note: required for `posts::columns::id` to implement `AppearsOnTable<FromClause<users::table>>`
  --> tests/fail/invalid_group_by.rs:15:9
   |
15 |         id -> Integer,
   |         ^^
   = note: associated types for the current `impl` cannot be restricted in `where` clauses
   = note: required for `SelectStatement<FromClause<users::table>, diesel::query_builder::select_clause::SelectClause<users::columns::id>>` to implement `GroupByDsl<posts::columns::id>`

error[E0271]: type mismatch resolving `<FromClause<Alias<user1>> as AppearsInFromClause<table>>::Count == Once`
  --> tests/fail/invalid_group_by.rs:43:10
   |
43 |         .group_by(posts::id)
   |          ^^^^^^^^ expected `Never`, found `Once`
   |
note: required for `posts::columns::id` to implement `AppearsOnTable<FromClause<Alias<user1>>>`
  --> tests/fail/invalid_group_by.rs:15:9
   |
15 |         id -> Integer,
   |         ^^
   = note: associated types for the current `impl` cannot be restricted in `where` clauses
   = note: required for `SelectStatement<FromClause<Alias<user1>>>` to implement `GroupByDsl<posts::columns::id>`

error[E0277]: the trait bound `AliasedField<user1, users::columns::id>: ValidGrouping<posts::columns::id>` is not satisfied
  --> tests/fail/invalid_group_by.rs:44:17
   |
44 |         .select(user_alias.field(users::id))
   |          ------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ValidGrouping<posts::columns::id>` is not implemented for `AliasedField<user1, users::columns::id>`, which is required by `SelectStatement<FromClause<Alias<user1>>, diesel::query_builder::select_clause::DefaultSelectClause<FromClause<Alias<user1>>>, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::NoWhereClause, diesel::query_builder::order_clause::NoOrderClause, LimitOffsetClause<NoLimitClause, NoOffsetClause>, diesel::query_builder::group_by_clause::GroupByClause<posts::columns::id>>: SelectDsl<_>`
   |          |
   |          required by a bound introduced by this call
   |
   = help: the following other types implement trait `ValidGrouping<GroupByClause>`:
             <AliasedField<S, C2> as ValidGrouping<AliasedField<S, C1>>>
             <AliasedField<S, C> as ValidGrouping<()>>
   = note: required for `SelectStatement<FromClause<Alias<user1>>, DefaultSelectClause<FromClause<...>>, ..., ..., ..., ..., ...>` to implement `SelectDsl<AliasedField<user1, users::columns::id>>`
note: required by a bound in `diesel::QueryDsl::select`
  --> $DIESEL/src/query_dsl/mod.rs
   |
   |     fn select<Selection>(self, selection: Selection) -> Select<Self, Selection>
   |        ------ required by a bound in this associated function
...
   |         Self: methods::SelectDsl<Selection>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `QueryDsl::select`
   = note: consider using `--verbose` to print the full type name to the console

error[E0271]: type mismatch resolving `<FromClause<table> as AppearsInFromClause<Alias<post1>>>::Count == Once`
  --> tests/fail/invalid_group_by.rs:49:10
   |
49 |         .group_by(post_alias.field(posts::id))
   |          ^^^^^^^^ expected `Never`, found `Once`
   |
   = note: required for `AliasedField<post1, posts::columns::id>` to implement `AppearsOnTable<FromClause<users::table>>`
   = note: required for `SelectStatement<FromClause<users::table>>` to implement `GroupByDsl<AliasedField<post1, posts::columns::id>>`

error[E0277]: the trait bound `AliasedField<post1, posts::columns::id>: IsContainedInGroupBy<users::columns::id>` is not satisfied
  --> tests/fail/invalid_group_by.rs:50:10
   |
50 |         .select(users::id)
   |          ^^^^^^ the trait `IsContainedInGroupBy<users::columns::id>` is not implemented for `AliasedField<post1, posts::columns::id>`, which is required by `SelectStatement<FromClause<users::table>, diesel::query_builder::select_clause::DefaultSelectClause<FromClause<users::table>>, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::NoWhereClause, diesel::query_builder::order_clause::NoOrderClause, LimitOffsetClause<NoLimitClause, NoOffsetClause>, diesel::query_builder::group_by_clause::GroupByClause<AliasedField<post1, posts::columns::id>>>: SelectDsl<_>`
   |
   = help: the following other types implement trait `IsContainedInGroupBy<T>`:
             <(T0, T1) as IsContainedInGroupBy<Col>>
             <(T0, T1, T2) as IsContainedInGroupBy<Col>>
             <(T0, T1, T2, T3) as IsContainedInGroupBy<Col>>
             <(T0, T1, T2, T3, T4) as IsContainedInGroupBy<Col>>
             <(T0, T1, T2, T3, T4, T5) as IsContainedInGroupBy<Col>>
             <(T0, T1, T2, T3, T4, T5, T6) as IsContainedInGroupBy<Col>>
             <(T0, T1, T2, T3, T4, T5, T6, T7) as IsContainedInGroupBy<Col>>
             <(T0, T1, T2, T3, T4, T5, T6, T7, T8) as IsContainedInGroupBy<Col>>
           and $N others
note: required for `users::columns::id` to implement `ValidGrouping<AliasedField<post1, posts::columns::id>>`
  --> tests/fail/invalid_group_by.rs:8:9
   |
8  |         id -> Integer,
   |         ^^
   = note: required for `SelectStatement<FromClause<table>, DefaultSelectClause<FromClause<table>>, ..., ..., ..., ..., ...>` to implement `SelectDsl<users::columns::id>`
   = note: consider using `--verbose` to print the full type name to the console

error[E0271]: type mismatch resolving `<FromClause<Alias<user1>> as AppearsInFromClause<Alias<post1>>>::Count == Once`
  --> tests/fail/invalid_group_by.rs:55:10
   |
55 |         .group_by(post_alias.field(posts::id))
   |          ^^^^^^^^ expected `Never`, found `Once`
   |
   = note: required for `AliasedField<post1, posts::columns::id>` to implement `AppearsOnTable<FromClause<Alias<user1>>>`
   = note: required for `SelectStatement<FromClause<Alias<user1>>>` to implement `GroupByDsl<AliasedField<post1, posts::columns::id>>`

error[E0277]: the trait bound `AliasedField<user1, users::columns::id>: ValidGrouping<AliasedField<post1, posts::columns::id>>` is not satisfied
  --> tests/fail/invalid_group_by.rs:56:17
   |
56 |         .select(user_alias.field(users::id))
   |          ------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ValidGrouping<AliasedField<post1, posts::columns::id>>` is not implemented for `AliasedField<user1, users::columns::id>`, which is required by `SelectStatement<FromClause<Alias<user1>>, diesel::query_builder::select_clause::DefaultSelectClause<FromClause<Alias<user1>>>, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::NoWhereClause, diesel::query_builder::order_clause::NoOrderClause, LimitOffsetClause<NoLimitClause, NoOffsetClause>, diesel::query_builder::group_by_clause::GroupByClause<AliasedField<post1, posts::columns::id>>>: SelectDsl<_>`
   |          |
   |          required by a bound introduced by this call
   |
   = help: the following other types implement trait `ValidGrouping<GroupByClause>`:
             <AliasedField<S, C2> as ValidGrouping<AliasedField<S, C1>>>
             <AliasedField<S, C> as ValidGrouping<()>>
   = note: required for `SelectStatement<FromClause<Alias<user1>>, DefaultSelectClause<FromClause<...>>, ..., ..., ..., ..., ...>` to implement `SelectDsl<AliasedField<user1, users::columns::id>>`
note: required by a bound in `diesel::QueryDsl::select`
  --> $DIESEL/src/query_dsl/mod.rs
   |
   |     fn select<Selection>(self, selection: Selection) -> Select<Self, Selection>
   |        ------ required by a bound in this associated function
...
   |         Self: methods::SelectDsl<Selection>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `QueryDsl::select`
   = note: consider using `--verbose` to print the full type name to the console

error[E0271]: type mismatch resolving `<FromClause<Alias<user1>> as AppearsInFromClause<table>>::Count == Once`
  --> tests/fail/invalid_group_by.rs:62:10
   |
62 |         .group_by(posts::id)
   |          ^^^^^^^^ expected `Never`, found `Once`
   |
note: required for `posts::columns::id` to implement `AppearsOnTable<FromClause<Alias<user1>>>`
  --> tests/fail/invalid_group_by.rs:15:9
   |
15 |         id -> Integer,
   |         ^^
   = note: associated types for the current `impl` cannot be restricted in `where` clauses
   = note: required for `SelectStatement<FromClause<Alias<user1>>, diesel::query_builder::select_clause::SelectClause<AliasedField<user1, users::columns::id>>>` to implement `GroupByDsl<posts::columns::id>`

error[E0271]: type mismatch resolving `<FromClause<table> as AppearsInFromClause<Alias<post1>>>::Count == Once`
  --> tests/fail/invalid_group_by.rs:68:10
   |
68 |         .group_by(post_alias.field(posts::id))
   |          ^^^^^^^^ expected `Never`, found `Once`
   |
   = note: required for `AliasedField<post1, posts::columns::id>` to implement `AppearsOnTable<FromClause<users::table>>`
   = note: required for `SelectStatement<FromClause<users::table>, diesel::query_builder::select_clause::SelectClause<users::columns::id>>` to implement `GroupByDsl<AliasedField<post1, posts::columns::id>>`

error[E0271]: type mismatch resolving `<FromClause<Alias<user1>> as AppearsInFromClause<Alias<post1>>>::Count == Once`
  --> tests/fail/invalid_group_by.rs:74:10
   |
74 |         .group_by(post_alias.field(posts::id))
   |          ^^^^^^^^ expected `Never`, found `Once`
   |
   = note: required for `AliasedField<post1, posts::columns::id>` to implement `AppearsOnTable<FromClause<Alias<user1>>>`
   = note: required for `SelectStatement<FromClause<Alias<user1>>, diesel::query_builder::select_clause::SelectClause<AliasedField<user1, users::columns::id>>>` to implement `GroupByDsl<AliasedField<post1, posts::columns::id>>`
